File: range-extraction.js
function solution(list) {
	return (
		list
			// get the differences between each element in the array.
			.map((n, i, arr) => {
				return { n, diff: (arr[i - 1] || 0) - n };
			})

			// produce an array split when the difference is greater that |1| and differs from the previous.
			.reduce(
				(ranges, n) => {
					const prev = ranges[0];
					if (
						n.diff ** 2 !== 1 &&
						prev[prev.length - 1] &&
						n.diff !== prev[prev.length - 1].diff
					) {
						ranges.unshift([n.n]);
					} else {
						prev.push(n.n);
					}
					return ranges;
				},
				[[]]
			)

			// reduce the array ranges.
			.map((range) =>
				range.length > 2
					? [range.shift(), range.pop()].join("-")
					: range.join(",")
			)

			// reverse the array, as we unshifted new values.
			.reverse()

			// join the array series.
			.join(",")
	);
}

File: range-extraction.py
def solution(args):
	diffs = [
		{
			'n': n, 
			'diff': (args[i-1 if i > 0 else 0] - n) if i > 0 else 0,
		} for i, n in enumerate(args)
	]

	ranges = [[]]
	for n in diffs:
		prev = ranges[-1]
		if n['diff'] ** 2 != 1 and prev and prev[-1] and n['diff'] != prev[-1]['diff']:
			ranges.append([n])
		else:
			prev.append(n)
	
	print(diffs, [n['n'] for n in range for range in ranges])